pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -907,6 +909,43 @@ func (c *Controller) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.Serv
 	return out, nil
 }
 
+func (c *Controller) hydrateForeignServiceInstance(si *model.ServiceInstance) ([]*model.ServiceInstance, error) {
+	out := []*model.ServiceInstance{}
+	// find the workload entry's service by label selector
+	// rather than scanning through our internal map of model.services, get the services via the k8s apis
+	dummyPod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{Namespace: si.Service.Attributes.Namespace, Labels: si.Endpoint.Labels},
+	}
```

Comment:
Why do we need to go through k8s for this? We have the label selector fields in our service object. We simply need to iterate through all the services in the namespace and pick matching ones.. Eliminates the need to create dummy pods or go through k8s apis

Reply:
This is more efficient. Our internal map is keyed off hostname, k8s is keyed off namespace. So we would have to search globally.

There is a lot of cleanup around this code in general we should do as a followup